google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryRequest.java

```diff
@@ -102,6 +106,33 @@ public Builder setQuery(String query) {
     }
 
     /**
+     * Adds a QueryParameter to the list of query parameters. See
+     * {@link #setQueryParameters(Iterable)} for more details on the input requirements.
+     */
+    public Builder addQueryParameter(QueryParameter queryParameter) {
+      if (queryParameters == null) {
+        queryParameters = Lists.newArrayList();
+      }
+      queryParameters.add(checkNotNull(queryParameter));
+      return this;
+    }
+
+    /**
+     * Sets the query parameters for the query.
+     *
+     * The set of query parameters must either be all positional or all named parameters.
+     * Positional parameters are denoted in the query with a question mark (?), and named
```

Comment:
We could enforce that during compilation by having separate classes  for positional and name (could inherit from the same base class if needed). Then we can have separate method for adding them.
Also, not compile time but the single case of addQueryParameter (which also should have 2 overloaded methods) could throw IllegalArgumentException when we already have non empty list of the other type (positional or named).